dj/sql/parsing/ast.py

```diff
@@ -53,56 +54,77 @@ class Node(ABC):
 
     DJ nodes are python dataclasses with the following patterns:
         - Attributes are either
-            - primitives (int, float, str, bool, None)
+            - PRIMITIVES (int, float, str, bool, None)
             - iterable from (list, tuple, set)
             - Enum
             - descendant of `Node`
         - Attributes starting with '_' are "obfuscated" and are not included in `children`
 
     """
 
-    _parents: Set["Node"]
+    parent: Optional["Node"] = None
+    parent_key: Optional[str] = None
 
     def __post_init__(self):
-        self._parents = set()
         self.add_self_as_parent()
 
-    @property
-    def parents(self) -> Set["Node"]:
-        """get the parents of the node"""
-        return self._parents
-
-    def clear_parents(self: TNode) -> TNode:
-        """remove all parents from the node"""
-        self._parents = set()
-        return self
-
-    def add_parents(self: TNode, *parents: "Node") -> TNode:
-        """add parents to the node"""
-        self.parents.update(parents)
+    def clear_parent(self: TNode) -> TNode:
+        """remove parent from the node"""
+        self.parent = None
         return self
 
-    def remove_parents(self: TNode, *parents: "Node") -> TNode:
-        """remove potential parents if they belong to the node"""
-        self._parents -= set(parents)
+    def set_parent(self: TNode, parent: "Node", parent_key: str) -> TNode:
+        """add parent to the node"""
+        self.parent = parent
+        self.parent_key = parent_key
         return self
 
     def add_self_as_parent(self: TNode) -> TNode:
         """adds self as a parent to all children"""
-        for child in self.children:
-            child.add_parents(self)
+        for name, child in self.fields(
+            flat=True,
+            nodes_only=True,
+            obfuscated=False,
+            nones=False,
+            named=True,
+        ):
+            child.set_parent(self, name)
         return self
 
+    def __setattr__(self, key: str, value: Any):
+        """Facilitates setting children using `.` syntax ensuring parent is attributed"""
+        if key == "parent":
+            object.__setattr__(self, key, value)
+            return
+
+        object.__setattr__(self, key, value)
+        for child in flatten(value):
+            if isinstance(child, Node) and not key.startswith("_"):
+                child.set_parent(self, key)
+
+    def get_nearest_parent_of_type(
+        self: "Node",
+        node_type: Type[TNode],
+    ) -> Optional[TNode]:
+        """traverse up the tree until you find a node of `node_type` or hit the root"""
```

Comment:
What's the purpose of this function? Why would you want to get a single node of some type? What if there are more nodes of the same type at the same distance?

Reply:
The intention of this function was that some times you might find yourself with a table or column that belongs to an ast, but you don't necessarily know where. Maybe it is in the uppermost query or maybe it is buried into some subquery so you might ask for the nearest select. 

I expected that this might be useful while building if we needed to add joins for dimensions you want to make sure you are adding them to the correct query, but it is definitely vestigial right now (no building in this PR) and actually in the build code I've got I just call build on the subqueries so I probably don't need this method 🙃

> What if there are more nodes of the same type at the same distance?

I don't think this is possible is it? Nodes currently only have a single parent, so as you trace up the AST towards the root you'll always encounter a unique node of a specific type.

Ought I just get rid of this?

Reply:
> I don't think this is possible is it? Nodes currently only have a single parent, so as you trace up the AST towards the root you'll always encounter a unique node of a specific type.

Oh... inside a query that is. Makes sense.

> Ought I just get rid of this?

If you are not using it, no need to keep it.

Reply:
Wait... nodes with a `JOIN` will have multiple parents.

Reply:
@betodealmeida "nodes" is being overloaded here. Nodes in the SQL AST only have single parents but DJ Nodes can definitely have multiple parents.

Reply:
Ah, sorry! I just got back home and I'm catching up. :)

Reply:
It's worth trying to come up with different names for these classes. When they both need to be imported to the same file it's confused me a couple of times and is solved by doing something like `from dj.sql.parsing.ast import Column as ASTColumn` which I don't think is the best solution. Any suggestions are welcome.

Reply:
They are naturally namespaced already, we could use it like

```python
from dj.sql.parsing import ast
...
ast.Column(ast.Name('x'))
```

Reply:
I agree this is a small issue. And prefixing it either during import or in the name looks much cleaner. The same problem exists for the Node classes. We should clearly make them distinct from the DJ dag objects. Otherwise confusion is guaranteed.